torch/lib/c10d/ProcessGroupRoundRobin.cpp

```diff
@@ -0,0 +1,102 @@
+  iterator_ = processGroups_.begin();
+}
+
+ProcessGroupRoundRobin::~ProcessGroupRoundRobin() {}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::broadcast(
+    std::vector<at::Tensor>& tensors,
+    const BroadcastOptions& opts) {
+  return next()->broadcast(tensors, opts);
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::allreduce(
+    std::vector<at::Tensor>& tensors,
+    const AllreduceOptions& opts) {
+  return next()->allreduce(tensors, opts);
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::allreduce_coalesced(
+    std::vector<at::Tensor>& tensors,
+    const AllreduceCoalescedOptions& opts) {
+  return next()->allreduce_coalesced(tensors, opts);
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::reduce(
+    std::vector<at::Tensor>& tensors,
+    const ReduceOptions& opts) {
+  return next()->reduce(tensors, opts);
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::allgather(
+    std::vector<std::vector<at::Tensor>>& outputs,
+    std::vector<at::Tensor>& inputs,
+    const AllgatherOptions& opts) {
+  return next()->allgather(outputs, inputs, opts);
+};
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::gather(
+    std::vector<std::vector<at::Tensor>>& outputs,
+    std::vector<at::Tensor>& inputs,
+    const GatherOptions& opts) {
+  return next()->gather(outputs, inputs, opts);
+};
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::scatter(
+    std::vector<at::Tensor>& outputs,
+    std::vector<std::vector<at::Tensor>>& inputs,
+    const ScatterOptions& opts) {
+  return next()->scatter(outputs, inputs, opts);
+};
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::reduce_scatter(
+    std::vector<at::Tensor>& outputs,
+    std::vector<std::vector<at::Tensor>>& inputs,
+    const ReduceScatterOptions& opts) {
+  return next()->reduce_scatter(outputs, inputs, opts);
+};
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::send(
+    std::vector<at::Tensor>& /* unused */,
+    int /* unused */,
+    int /* unused */) {
+  throw std::runtime_error("ProcessGroupRoundRobin does not support send");
+};
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::recv(
+    std::vector<at::Tensor>& /* unused */,
+    int /* unused */,
+    int /* unused */) {
+  throw std::runtime_error("ProcessGroupRoundRobin does not support recv");
+};
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::recvAnysource(
+    std::vector<at::Tensor>& /* unused */,
+    int /* unused */) {
+  throw std::runtime_error("ProcessGroupRoundRobin does not support recv");
+};
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::barrier(
+    const BarrierOptions& /* unused */) {
+  throw std::runtime_error("ProcessGroupRoundRobin does not support barrier");
+};
+
+const std::shared_ptr<ProcessGroup>& ProcessGroupRoundRobin::next() {
+  auto& processGroup = *iterator_;
```

Comment:
do we need lock?

Reply:
No lock needed. All process groups can only be used by a single thread anyway (because the ordering of the calls must be globally deterministic).

Reply:
> All process groups can only be used by a single thread anyway

Does this still hold if we use tags? IIUC, we do support using ProcessGroup with different tag in different thread, do we want the same thing for ProcessGroupRoundRobin as well? If no, let's explicitly say that in the comments/docs.

Reply:
We only support that by accident, for send/recv only on ProcessGroupGloo. In general, collective operations must be issued in globally deterministic order, and any races here would immediately lead to incorrectness. Adding a mutex here would give a false sense of safety, where in reality there is none.

For example, if you have two threads T1 and T2, and both want to execute an allreduce. T1 operates on 1 float and T2 operates on 10 floats. Now consider two processes that do this. There is nothing we can do to ensure the correct ordering of these operations if the application code doesn't already take care of it. Without the application doing anything, you could easily end up with T1 trying to work with T2 on the other process, and causing a crash/

Reply:
Just checked and the support in ProcessGroupGloo is accidental. There are no facilities for thread safety there at all. That it works in the first place is just because send/recv pretty much just patching through to Gloo directly.